src/lib/coreaudio/CoreAudioHelpers.cxx

```diff
@@ -0,0 +1,261 @@
+	if (isprint(buffer[1]) && isprint(buffer[2]) && isprint(buffer[3]) && isprint(buffer[4]))
+	{
+		buffer[0] = buffer[5] = '\'';
+		buffer[6] = '\0';
+	}
+	else
+	{
+		// no, format it as an integer
+		sprintf(buffer, "%d", (int)error);
+	}
+	return std::string(buffer);
+}
+
+const char *
+StreamDescriptionToString(const AudioStreamBasicDescription desc, std::string &str)
+{
+	char fourCC[5] = {
+		(char)((desc.mFormatID >> 24) & 0xFF),
+		(char)((desc.mFormatID >> 16) & 0xFF),
+		(char)((desc.mFormatID >>  8) & 0xFF),
+		(char) (desc.mFormatID        & 0xFF),
+		0
+	};
+
+	std::stringstream sstr;
+	switch (desc.mFormatID)
+	{
+		case kAudioFormatLinearPCM:
+			sstr  << "["
+				<< fourCC
+				<< "] "
+				<< ((desc.mFormatFlags & kAudioFormatFlagIsNonMixable) ? "" : "Mixable " )
+				<< ((desc.mFormatFlags & kAudioFormatFlagIsNonInterleaved) ? "Non-" : "" )
+				<< "Interleaved "
+				<< desc.mChannelsPerFrame
+				<< " Channel "
+				<< desc.mBitsPerChannel
+				<< "-bit "
+				<< ((desc.mFormatFlags & kAudioFormatFlagIsFloat) ? "Floating Point " : "Signed Integer ")
+				<< ((desc.mFormatFlags & kAudioFormatFlagIsBigEndian) ? "BE" : "LE")
+				<< " ("
+				<< (UInt32)desc.mSampleRate
+				<< "Hz)";
+			str = sstr.str();
+			break;
+		case kAudioFormatAC3:
+			sstr  << "["
+				<< fourCC
+				<< "] "
+				<< ((desc.mFormatFlags & kAudioFormatFlagIsBigEndian) ? "BE" : "LE")
+				<< " AC-3/DTS ("
+				<< (UInt32)desc.mSampleRate
+				<< "Hz)";
+			str = sstr.str();
+			break;
+		case kAudioFormat60958AC3:
+			sstr  << "["
+				<< fourCC
+				<< "] AC-3/DTS for S/PDIF "
+				<< ((desc.mFormatFlags & kAudioFormatFlagIsBigEndian) ? "BE" : "LE")
+				<< " ("
+				<< (UInt32)desc.mSampleRate
+				<< "Hz)";
+			str = sstr.str();
+			break;
+		default:
+			sstr  << "["
+				<< fourCC
+				<< "]";
+			break;
+	}
+	return str.c_str();
+}
+
+AudioBufferList *
+AllocateABL(const AudioStreamBasicDescription asbd, const UInt32 capacity_frames)
+{
+	AudioBufferList *buffer_list = nullptr;
+	UInt32 num_buffers = (asbd.mFormatFlags & kAudioFormatFlagIsNonInterleaved) ? asbd.mChannelsPerFrame : 1;
+	
+	if((buffer_list = static_cast<AudioBufferList *>(calloc(1, offsetof(AudioBufferList, mBuffers) + (sizeof(AudioBuffer) * num_buffers)))) == nullptr)
+		throw std::runtime_error("Unable to allocate memory for AudioBufferList.");
+	else
+	{
+		buffer_list->mNumberBuffers = num_buffers;
+		for(UInt32 buffer_index = 0; buffer_index < buffer_list->mNumberBuffers; ++buffer_index)
+		{
+			try
+			{
+				AllocateAudioBuffer(buffer_list->mBuffers[buffer_index], asbd, capacity_frames);
+			}
+			catch (...)
+			{
+				DeallocateABL(buffer_list);
+				std::throw_with_nested("Unable to allocate memory for AudioBufferList.");
+			}
+		}
+	}
+	return buffer_list;
+}
+
+void
+DeallocateABL(AudioBufferList *buffer_list)
+{
+	if(buffer_list != nullptr)
+	{
+		for(UInt32 buffer_index = 0; buffer_index < buffer_list->mNumberBuffers; ++buffer_index) {
```

Comment:
Why `UInt32`? Forcing the compiler into one register size may reduce performance.

Reply:
for loops changed to unsigned int